### DISCUSSION TITLE
Manipulating JSON?

### DISCUSSION THREAD

**Discussion by Yoric** [diesel-rs/diesel#2888] [Q&A]
I have the following postgres statement:

```sql
INSERT INTO destination (a, b, c, d)
SELECT data->>'a', b, data->>'c', data->'d'
FROM source WHERE label = 'abcd' AND data->>'type' = 'foobar;
```
where `data` is a `JSONB`.

For people not familiar with Postgres' handling of JSON, I'm:
- copying strings (basically `data["a"].as_str()`, `data["b"].as_str()`, `data["c"].as_str()`);
- copying sub-objects (`data["d"]`)
- filtering based on the content of a JSON object.

For the time being, I'm using `sql_query`. Is there a more strongly-typed manner of implementing this with diesel? I realize that Postgres doesn't support any kind of scheme for the JSON, but it might be possible to make *some* of it strongly-typed.

If not, is this somewhere on the roadmap?

**Comment by Yoric** [comment]
The following seems to work:

```rust
schema::source::dsl::source
  .select((sql("data->>'a"), schema::destination::dsl::b, sql("data->>'c"), sql("data->'d")))
  .filter(schema::source::dsl::label.eq("abcd"))
  .filter(sql("data->>'type' = 'foobar'"))
  .insert_into(schema::destination::dsl::destination)
  .execute(&conn);
```

Is there a better way?

**Comment by weiznich** [comment]
As you already found out insert from select is supported by diesels built in query dsl. Json operators are not provided yet, but they can be defined using the [`diesel_infix_operator!`]( https://docs.diesel.rs/1.4.x/diesel/macro.diesel_infix_operator.html) macro. See #44 for more details. 
That written I would likely accept a PR that adds those operators to diesel itself as long as it contains tests + documentation.

**Comment by AlbertMarashi** [comment]
https://docs.diesel.rs/2.0.x/diesel/expression_methods/trait.PgJsonbExpressionMethods.html#method.contains

Came here from google looking to find how to filter `Jsonb` data. So here's how I did it
```rs
.filter(application.contains(&serde_json::json!({ "user_id": self.id })))
```
